src/TextInputWidget.cc

```diff
@@ -101,6 +105,39 @@ FilteredTextEdit::keyPressEvent(QKeyEvent *event)
         }
 }
 
+bool
+FilteredTextEdit::canInsertFromMimeData(const QMimeData *source) const
+{
+        return (source->hasImage() || source->hasText() ||
+                QTextEdit::canInsertFromMimeData(source));
+}
+
+void
+FilteredTextEdit::insertFromMimeData(const QMimeData *source)
+{
+        if (source->hasImage()) {
+                QPixmap const img = qvariant_cast<QPixmap>(source->imageData());
+                previewDialog_    = new dialogs::PreviewImageOverlay{img, this};
```

Comment:
When does the `PreviewImageOverlay` you create here get deleted? Why does it need to be dynamically allocated at all?

Reply:
The reason it is dynamically allocated is because the dialog is created on a paste event, which is dynamic behaviour by nature. The same approach is used in [`src/SideBarActions.cc`](https://github.com/mujx/nheko/blob/master/src/SideBarActions.cc#L39) and [`src/UserInfoWidget.cc`](https://github.com/mujx/nheko/blob/master/src/UserInfoWidget.cc#L95), to name a few examples.

How it is deleted is an oversight. I have locally changed `previewDialog_` to be a `QSharedPointer` so that it can be de-allocated automatically. I also pass `this` as a parameter to the constructor as to create a parent-child relationship with Qt's object model.

Please let me know if you have a better approach.

Reply:
The UI semantics you describe are orthogonal to the actual construction/destruction of the object; you could let `PreviewImageOverlay` be a regular class member, and just show/hide/reset it as necessary. Dynamic allocation can be simpler, but requires care to avoid introducing leaks. Here, using Qt's parent-child relationship does not avoid a leak because many pastes may occur in the lifetime of a `TextInputWidget`.

Switching to a `QSharedPointer` fixes the problem, but may be confusing, as it's presumably not intended that ownership of the value be shared. Instead, I'd suggest using `std::unique_ptr`, and furthermore replacing the call to `new` with `std::make_unique`; calling `new` explicitly should generally be avoided, as it's prone to this type of mistake.